master/buildbot/worker/latent.py

```diff
@@ -97,6 +100,15 @@ def stop_instance(self, fast=False):
     def substantiated(self):
         return self.conn is not None
 
+    def retry_check(self):
+        # responsible for checking if the build should continue to retry
+        # or stop trying to substantiate
+        self.retry_count += 1
+        return self.retry_count < self.max_retry_count
```

Comment:
we need to make sure that this retry counter is from the same build! Here you are a bit re-implementing the quarantine timer